mops/src/sasaw/cpu.tpp

```diff
@@ -18,7 +19,65 @@ void mops::sparse_accumulation_scatter_add_with_weights(
     Tensor<int32_t, 1> indices_output_1,
     Tensor<int32_t, 1> indices_output_2
 ) {
-    // TODO
+    check_sizes(A, "A", 0, B, "B", 0, "sasaw");
+    check_sizes(W, "W", 0, output, "output", 0, "sasaw");
+    check_sizes(B, "B", 1, W, "W", 2, "sasaw");
+    check_sizes(C, "C", 0, indices_A, "indices_A", 0, "sasaw");
+    check_sizes(C, "C", 0, indices_W_2, "indices_W_2", 0, "sasaw");
+    check_sizes(C, "C", 0, indices_output_2, "indices_output_2", 0, "sasaw");
+    check_sizes(A, "A", 0, indices_output_1, "indices_output_1", 0, "sasaw");
+    check_sizes(A, "A", 0, indices_W_1, "indices_W_1", 0, "sasaw");
+    check_index_tensor(indices_output_1, "indices_output_1", output.shape[0], "sasaw");
+    check_index_tensor(indices_W_1, "indices_W_1", output.shape[0], "sasaw");
+    check_index_tensor(indices_A, "indices_A", A.shape[1], "sasaw");
+    check_index_tensor(indices_W_2, "indices_W_2", B.shape[1], "sasaw");
+    check_index_tensor(indices_output_2, "indices_output_2", output.shape[1], "sasaw");
+
+    scalar_t* o_ptr = output.data;
+    scalar_t* a_ptr = A.data;
+    scalar_t* b_ptr = B.data;
+    scalar_t* w_ptr = W.data;
+    scalar_t* c_ptr = C.data;
+    int* i_ptr = indices_output_1.data;
+    int* j_ptr = indices_W_1.data;
+    int* m_1_ptr = indices_A.data;
+    int* m_2_ptr = indices_W_2.data;
+    int* m_3_ptr = indices_output_2.data;
+
+    size_t N = C.shape[0];
+    size_t size_a = A.shape[1];
+    size_t size_b = B.shape[1];
+    size_t size_output_first_dim = output.shape[0];
+    size_t output_shift_first_dim = output.shape[1] * output.shape[2];
+    size_t w_shift_first_dim = W.shape[1] * W.shape[2];
+    size_t output_shift_second_dim = output.shape[2];
+    size_t w_shift_second_dim = W.shape[2];
+
+    // For each index in the first dimension of the outputs,
+    // get what indices in the inputs should write to it
+    std::vector<std::vector<size_t>> write_list = get_write_list(indices_output_1);
+    std::fill(o_ptr, o_ptr+output.shape[0]*output_shift_first_dim, static_cast<scalar_t>(0.0));
+
+    #pragma omp parallel for
+    for (size_t i = 0; i < size_output_first_dim; i++) {
+        // iterate over input indices that will write to the output index i
+        for (size_t e : write_list[i]) {
```

Comment:
I'm a bit lost as to what `e` means here. Maybe adding a comment or using a longer name would help.